src/Components/Server/test/Circuits/RemoteRendererTest.cs

```diff
@@ -88,22 +93,189 @@ public async Task ProcessBufferedRenderBatches_WritesRenders()
+            var renderIds = new List<long>();
+
+            var firstBatchTCS = new TaskCompletionSource<object>();
+            var secondBatchTCS = new TaskCompletionSource<object>();
+            var thirdBatchTCS = new TaskCompletionSource<object>();
+
+            var initialClient = new Mock<IClientProxy>();
+            initialClient.Setup(c => c.SendCoreAsync(It.IsAny<string>(), It.IsAny<object[]>(), It.IsAny<CancellationToken>()))
+                .Callback((string name, object[] value, CancellationToken token) =>
+                {
+                    renderIds.Add((long)value[1]);
+                })
+                .Returns(firstBatchTCS.Task);
+            var circuitClient = new CircuitClientProxy(initialClient.Object, "connection0");
+            var renderer = GetRemoteRenderer(serviceProvider, circuitClient);
+            var component = new TestComponent(builder =>
+            {
+                builder.OpenElement(0, "my element");
+                builder.AddContent(1, "some text");
+                builder.CloseElement();
+            });
+
+            var client = new Mock<IClientProxy>();
+            client.Setup(c => c.SendCoreAsync(It.IsAny<string>(), It.IsAny<object[]>(), It.IsAny<CancellationToken>()))
+                .Callback((string name, object[] value, CancellationToken token) =>
+                {
+                    renderIds.Add((long)value[1]);
+                })
+                .Returns<string, object[], CancellationToken>((n, v, t) => (long)v[1] == 3 ? secondBatchTCS.Task : thirdBatchTCS.Task);
+
+            var componentId = renderer.AssignRootComponentId(component);
+            component.TriggerRender();
+            renderer.OnRenderCompleted(2, null);
+            firstBatchTCS.SetResult(null);
+
+            circuitClient.SetDisconnected();
+            component.TriggerRender();
+            component.TriggerRender();
+
+            // Act
+            circuitClient.Transfer(client.Object, "new-connection");
+            var task = renderer.ProcessBufferedRenderBatches();
+
+            renderer.OnRenderCompleted(3, null);
+            secondBatchTCS.SetResult(null);
+
+            renderer.OnRenderCompleted(2, null);
 
             // Assert
-            client.Verify(c => c.SendCoreAsync("JS.RenderBatch", It.IsAny<object[]>(), It.IsAny<CancellationToken>()), Times.Exactly(2));
+            Assert.Equal(new long[] { 2, 3, 4 }, renderIds);
+            var pendingBatch = Assert.Single(renderer.PendingRenderBatches);
+            Assert.Equal(4, pendingBatch.BatchId);
+            renderer.OnRenderCompleted(4, null);
+            thirdBatchTCS.SetResult(null);
+            Assert.Empty(renderer.PendingRenderBatches);
+
+            Assert.True(task.Wait(3000), "One or more render batches werent acknowledged");
+
+            await task;
+        }
+
+        [Fact]
+        public async Task GracefullyRecoversFromMissingClientAcknowledge()
```

Comment:
Unless there's a legitimate scenario for this, I propose we change this to "throws and kills connection if client ACKs are received out of order".

Reply:
Changed it to throw